consensus/hotstuff/safetyrules/safety_rules_test.go

```diff
@@ -1,165 +1,526 @@
 
-	_, err := voter.ProduceVote(block, curView)
-	require.Error(t, err)
-	require.Contains(t, err.Error(), "not safe")
-	require.True(t, model.IsNoVoteError(err))
+	expectedSafetyData := &hotstuff.SafetyData{
+		LockedOneChainView:      s.proposal.Block.QC.View,
+		HighestAcknowledgedView: s.proposal.Block.View,
+	}
+
+	expectedVote := makeVote(s.proposal.Block)
+	s.signer.On("CreateVote", s.proposal.Block).Return(expectedVote, nil).Once()
+	s.persister.On("PutSafetyData", expectedSafetyData).Return(nil).Once()
+
+	vote, err := s.safety.ProduceVote(s.proposal, s.proposal.Block.View)
+	require.NoError(s.T(), err)
+	require.NotNil(s.T(), vote)
+	require.Equal(s.T(), expectedVote, vote)
+	s.signer.AssertExpectations(s.T())
+	s.persister.AssertCalled(s.T(), "PutSafetyData", expectedSafetyData)
 }
 
-func testBelowVote(t *testing.T) {
-	// curView < blockView
-	blockView, curView, lastVotedView, isBlockSafe, isCommitteeMember := uint64(3), uint64(2), uint64(2), true, true
+// TestProduceVote_InvalidCurrentView tests that no vote is created if proposal is for invalid view
+func (s *SafetyRulesTestSuite) TestProduceVote_InvalidCurrentView() {
+	vote, err := s.safety.ProduceVote(s.proposal, s.proposal.Block.View+1)
+	require.Nil(s.T(), vote)
+	require.Error(s.T(), err)
+	s.persister.AssertNotCalled(s.T(), "PutSafetyData")
+}
 
-	// create voter
-	block, _, voter := createVoter(blockView, lastVotedView, isBlockSafe, isCommitteeMember)
+// TestProduceVote_NodeEjected tests that no vote is created if block proposer is ejected
+func (s *SafetyRulesTestSuite) TestProduceVote_ProposerEjected() {
+	*s.committee = mocks.DynamicCommittee{}
+	s.committee.On("IdentityByBlock", s.proposal.Block.BlockID, s.proposal.Block.ProposerID).Return(nil, model.NewInvalidSignerErrorf("node-ejected")).Once()
 
-	_, err := voter.ProduceVote(block, curView)
-	require.Error(t, err)
-	require.Contains(t, err.Error(), "expecting block for current view")
-	require.False(t, model.IsNoVoteError(err))
+	vote, err := s.safety.ProduceVote(s.proposal, s.proposal.Block.View)
+	require.Nil(s.T(), vote)
+	require.True(s.T(), model.IsNoVoteError(err))
+	s.persister.AssertNotCalled(s.T(), "PutSafetyData")
 }
 
-func testAboveVote(t *testing.T) {
-	// curView > blockView
-	blockView, curView, lastVotedView, isBlockSafe, isCommitteeMember := uint64(3), uint64(4), uint64(2), true, true
+// TestProduceVote_InvalidProposerIdentity tests that no vote is created if there was an exception retrieving proposer identity
+func (s *SafetyRulesTestSuite) TestProduceVote_InvalidProposerIdentity() {
+	*s.committee = mocks.DynamicCommittee{}
+	exception := errors.New("invalid-signer-identity")
+	s.committee.On("IdentityByBlock", s.proposal.Block.BlockID, s.proposal.Block.ProposerID).Return(nil, exception).Once()
+
+	vote, err := s.safety.ProduceVote(s.proposal, s.proposal.Block.View)
+	require.Nil(s.T(), vote)
+	require.ErrorIs(s.T(), err, exception)
```

Comment:
If `SafetyRules` had a bug and erroneously wrapped the exception into a `NotVoteError` this test would still pass. I think it is important to test this here:
```suggestion
	require.ErrorIs(s.T(), err, exception)
	require.False(s.T(), model.IsNoVoteError(err))
```

Would suggest to include a clarifying comment in the test's goDoc:
> We are specifically testing that unexpected errors are handled correctly, i.e. 
> that SafetyRules does _not_ erroneously wrap unexpected exceptions into the expected NoVoteError.